charms/worker/k8s/src/charm.py

```diff
@@ -1340,6 +1358,43 @@ def _on_refresh_certs_action(self, event: ops.ActionEvent):
             except (InvalidResponseError, K8sdConnectionError) as e:
                 event.fail(f"Failed to refresh certificates: {e}")
 
+    @on_error(
+        ops.BlockedStatus("Cannot change bootstrap config after bootstrap. Check logs."),
+        BootstrapConfigChangeError,
+    )
+    @on_error(
+        ops.ErrorStatus("Failed to get communicate with k8sd."),
+        InvalidResponseError,
+        K8sdConnectionError,
+    )
+    def _prevent_bootstrap_config_change(self):
+        """Prevent bootstrap config changes after bootstrap."""
+        log.info("Preventing bootstrap config changes after bootstrap")
+
+        if self.is_control_plane:
+            try:
+                ref_config = BootstrapConfigOptions.build(
+                    cluster_config=self.api_manager.get_cluster_config().metadata,
+                    node_status=self.api_manager.get_node_status().metadata,
+                )
+            except InvalidResponseError as e:
+                if e.code == http.HTTPStatus.SERVICE_UNAVAILABLE:
+                    log.info("k8sd is not ready, skipping bootstrap config check")
+                    return
+                raise
+        else:
+            try:
+                ref_config = BootstrapConfigOptions.build(
+                    node_status=self.api_manager.get_node_status().metadata,
+                )
+            except InvalidResponseError as e:
+                if e.code == http.HTTPStatus.SERVICE_UNAVAILABLE:
+                    log.info("k8sd is not ready, skipping bootstrap config check")
+                    return
+                raise
```

Comment:
Why do we need to catch the exception? Just for the info level log.  The reconciler will log for us already when the exception is raised. 

Are you saying that if the `build(...)` method really can raise an error and everything is fine and the reconciler should march on?  i notice the `e.code == http.HTTPStatus.SERVICE_UNAVAILABLE`.  What's this situation where k8ds is just kinda not ready and we're ok skipping the bootstrap config check.  I'd rather have the thing fail

```suggestion
        ref_config = BootstrapConfigOptions.build(
            cluster_config=(
                self.api_manager.get_cluster_config().metadata
                if self.is_control_plane
                else None
            ),
            node_status=self.api_manager.get_node_status().metadata,
        )
```

Reply:
I think it can happen when a cluster is not bootstrapped. Maybe we can play around with the ordering of the methods we call in the `_reconcile`.

Reply:
well, can we at least collapse this  `if...try...except...else...try...except` down to `try...except`


```suggestion
        try:
            ref_config = BootstrapConfigOptions.build(
                cluster_config=self.is_control_plane and self.api_manager.get_cluster_config().metadata,
                node_status=self.api_manager.get_node_status().metadata,
            )
        except InvalidResponseError as e:
            if e.code == http.HTTPStatus.SERVICE_UNAVAILABLE:
                log.info("k8sd is not ready, skipping bootstrap config check")
                return
            raise
```

Reply:
Great suggestion. It just needs some small changes tho, because `get_node_status` is restricted for worker nodes.
Changed it to:
```python
        try:
            ref_config = BootstrapConfigOptions.build(
                node_status=self.api_manager.get_node_status().metadata,
                cluster_config=self.api_manager.get_cluster_config().metadata
                if self.is_control_plane
                else None,
            )
        except InvalidResponseError as e:
            if e.code == http.HTTPStatus.SERVICE_UNAVAILABLE:
                log.info("k8sd is not ready, skipping bootstrap config check")
                return
            raise
```